src/index.ts

```diff
@@ -1,126 +1,46 @@
-    throw "compile() received neither an Array nor a String for its sources argument.";
-  }
+import { SpawnOptions, ChildProcess } from "child_process";
 
-  return typeof sources === "string" ? [sources] : sources;
-}
-
-function prepareOptions(options, spawnFn) {
-  return _.defaults({ spawn: spawnFn }, options, defaultOptions);
-}
-
-function prepareProcessArgs(sources, options) {
-  var preparedSources = prepareSources(sources);
-  var compilerArgs = compilerArgsFromOptions(options);
-
-  return ["make"].concat(preparedSources ? preparedSources.concat(compilerArgs) : compilerArgs);
-}
-
-function prepareProcessOpts(options) {
-  var env = _.merge({ LANG: 'en_US.UTF-8' }, process.env);
-  return _.merge({ env: env, stdio: "inherit", cwd: options.cwd }, options.processOpts);
-
-}
-
-function runCompiler(sources, options, pathToElm) {
-  if (typeof options.spawn !== "function") {
-    throw "options.spawn was a(n) " + (typeof options.spawn) + " instead of a function.";
-  }
-
-  var processArgs = prepareProcessArgs(sources, options);
-  var processOpts = prepareProcessOpts(options);
-
-  if (options.verbose) {
-    console.log(["Running", pathToElm].concat(processArgs).join(" "));
-  }
-
-  return options.spawn(pathToElm, processArgs, processOpts);
-}
+var elmBinaryName = "elm";
 
-function compilerErrorToString(err, pathToElm) {
-  if ((typeof err === "object") && (typeof err.code === "string")) {
-    switch (err.code) {
-      case "ENOENT":
-        return "Could not find Elm compiler \"" + pathToElm + "\". Is it installed?";
+temp.track();
 
-      case "EACCES":
-        return "Elm compiler \"" + pathToElm + "\" did not have permission to run. Do you need to give it executable permissions?";
 
-      default:
-        return "Error attempting to run Elm compiler \"" + pathToElm + "\":\n" + err;
-    }
-  } else if ((typeof err === "object") && (typeof err.message === "string")) {
-    return JSON.stringify(err.message);
-  } else {
-    return "Exception thrown when attempting to run Elm compiler " + JSON.stringify(pathToElm);
-  }
-}
-
-function compileSync(sources, options) {
-  var optionsWithDefaults = prepareOptions(options, options.spawn || spawn.sync);
+export function compile(sources: string | string[], options: Partial<Options>): ChildProcess {
+  var optionsWithDefaults = prepareOptions(options, spawn);
   var pathToElm = options.pathToElm || elmBinaryName;
 
   try {
-    return runCompiler(sources, optionsWithDefaults, pathToElm);
+    return runCompiler(sources, optionsWithDefaults, pathToElm)
+      .on('error', function (err) { throw (err); });
   } catch (err) {
     throw compilerErrorToString(err, pathToElm);
   }
 }
 
-function compile(sources, options) {
-  var optionsWithDefaults = prepareOptions(options, options.spawn || spawn);
+function compileSync(sources: string | string[], options: Partial<Options>): ChildProcess {
+  var optionsWithDefaults = prepareOptions(options, spawn.sync as any);
```

Comment:
The `Options` contain the spawn constructor. This is difficult, because the async versions use a different one than the sync versions, which I couldn't figure out how to type.

My suggestion would be to factor out the spawn constructor from the options and pass it separately.

Reply:
👍 we can definitely do some follow up PRs

Reply:
Yeah, you're right. It's probably best to do that in a separate PR, since it looks like a heavier refactoring.

Reply:
I've started such a refactor on https://github.com/Y0hy0h/node-elm-compiler/tree/refactor.